mm2src/coins/tendermint/tendermint_coin.rs

```diff
@@ -534,8 +552,26 @@ impl TendermintCoin {
             let to_address =
                 AccountId::from_str(&req.to).map_to_mm(|e| WithdrawError::InvalidAddress(e.to_string()))?;
 
+            let (account_id, priv_key) = match req.from {
+                Some(WithdrawFrom::HDWalletAddress(ref path_to_address)) => {
+                    let priv_key = coin
+                        .priv_key_policy
+                        .hd_wallet_derived_priv_key_or_err(path_to_address)?;
+                    let account_id = account_id_from_privkey(priv_key.as_slice(), &coin.account_prefix)
+                        .map_err(|e| WithdrawError::InternalError(e.to_string()))?;
+                    (account_id, priv_key)
+                },
+                Some(WithdrawFrom::AddressId(_)) | Some(WithdrawFrom::DerivationPath { .. }) => {
+                    return MmError::err(WithdrawError::UnexpectedFromAddress(
+                        "Withdraw from 'AddressId' or 'DerivationPath' is not supported yet for Tendermint!"
+                            .to_string(),
+                    ))
+                },
+                None => (coin.account_id.clone(), *coin.priv_key_policy.activated_key_or_err()?),
+            };
```

Comment:
1- Can we handle this in a more generic way; adding trait functions through `lp_coins` module? If we decide to go even further, we can use 1 function for all platforms(I guess this would take too much time for this PR)

2- We may need to save the generated addresses in coins context, so we don't generate the same addresses over and over for each request.

3- This block of code seems like duplicated a lot, can we replace them with a helper function?

Reply:
> 2- We may need to save the generated addresses in coins context, so we don't generate the same addresses over and over for each request.

I am actually thinking of using this https://github.com/KomodoPlatform/komodo-defi-framework/blob/afd268e3cc5997815875f36b40c14d6ca82e8534/mm2src/coins/utxo.rs#L588 and adding it to all coins in next PRs, It's necessary for transaction history and addresses balances. Only withdrawal from addresses in address cache will be allowed https://github.com/KomodoPlatform/komodo-defi-framework/blob/afd268e3cc5997815875f36b40c14d6ca82e8534/mm2src/coins/utxo.rs#L1532
`UtxoHDWallet` is only used for `Trezor` for now. I want to include the keypair instead of public here https://github.com/KomodoPlatform/komodo-defi-framework/blob/afd268e3cc5997815875f36b40c14d6ca82e8534/mm2src/coins/utxo.rs#L139 but this can't work for Trezor so I will keep `priv_key_policy` https://github.com/KomodoPlatform/komodo-defi-framework/blob/afd268e3cc5997815875f36b40c14d6ca82e8534/mm2src/coins/utxo.rs#L586 so to answer your first and third point.

> 1- Can we handle this in a more generic way; adding trait functions through lp_coins module? If we decide to go even further, we can use 1 function for all platforms(I guess this would take too much time for this PR)
>
> 3- This block of code seems like duplicated a lot, can we replace them with a helper function? By thinking quickly, maybe WithdrawFrom can implement helper function that can replace this and other duplications?

`WithdrawFrom::HDWalletAddress` will probably be removed and I will reuse `WithdrawFrom::AddressId`, `WithdrawFrom::DerivationPath` and this block of code will be changed/refactored.

Reply:
I am a bit confused with HDWalletAddress and DerivationPath variants. From the issue description I understand we were not satisfied with the old DerivationPath impl so added HDWalletAddress which better conforms to bip44. But from derive_secp256k1_secret fn I can see that both variants are used together. Wouldn't be better to have a self-contained implementation?

Reply:
> I am a bit confused with HDWalletAddress and DerivationPath variants. From the issue description I understand we were not satisfied with the old DerivationPath impl so added HDWalletAddress which better conforms to bip44.

`AddressId ` and `DerivationPath` were used for Trezor withdraw only before this PR. I added a new variant because chain below is named change for HD wallet standard. https://github.com/KomodoPlatform/komodo-defi-framework/blob/53d5e747020a2dfff1a0ed2bb2552b8adf30cfe0/mm2src/coins/hd_wallet.rs#L217
And also because Trezor uses a different function for withdraw. As I said above I will probably just reuse `AddressId` in next PRs.

Previous implementation is very confusing and requires a lot of refactoring, but I think the idea for `DerivationPath` is that in the request you can use the full derivation path string instead of the `HDAccountAddressId` struct since some prefer to provide the full derivation path e.g. `m/44'/60'/0'/0/0`

> But from derive_secp256k1_secret fn I can see that both variants are used together.

`derivation_path` in `derive_secp256k1_secret` `StandardHDPathToCoin` which make the name `derivation_path` not correct. I think it was named that since it's the name used in coin configs, changing this will need changing in coins repo.

Reply:
> Can you please elaborate more on what do you mean by self-contained implementation here?
what I meant was this code:
```
pub fn derive_secp256k1_secret(
    bip39_secp_priv_key: ExtendedPrivateKey<secp256k1::SecretKey>,
    derivation_path: &StandardHDPathToCoin,
    path_to_address: &StandardHDCoinAddress,
) -> MmResult<Secp256k1Secret, Bip32Error> {
    let mut account_der_path = derivation_path.to_derivation_path();
    account_der_path.push(ChildNumber::new(path_to_address.account, HARDENED).unwrap());
    account_der_path.push(ChildNumber::new(path_to_address.is_change as u32, NON_HARDENED).unwrap());
    account_der_path.push(ChildNumber::new(path_to_address.address_index, NON_HARDENED).unwrap());
```
it seems to me both params with paths are combined here, so path_to_address has only partial path. If I developed this I would use full path - that what I meant by self-contained Used standard path and just validated it for correctness.
F.e. as I can see bitcoind just uses unified standard full path (named hdKeypath) what is pretty clear to me. But it looks like we use different objects for pretty much the same concept. But maybe I do not know the history of this problem, sorry.
However if we eventually move to the single object AddressId, it would be nice (btw I would prefer it called HDKeyPath or similar).

Reply:
Part of the path comes from coin config (the fixed part) https://github.com/KomodoPlatform/coins/blob/f35dad48da6fe6cccafd9b50a4cf36492baa2744/coins#L6478 and the other part from API requests (the changing part related to `account'/change/address_index`) if we combine it before using this function we will have this part duplicated. I added this whole discussion to the issue checklist to not forget about it though https://github.com/KomodoPlatform/komodo-defi-framework/issues/1838#issuecomment-1675467570, this way I can look for ways for further optimization while working on next PRs.